### PR TITLE
Reland "dart:ui conversion from native to FfiNative" (#33116)"

### PR DESCRIPTION
**Pull request by dnfield**
This reverts commit 6f19a3e6cad4fe254dfc354a2747b3485b0afbc0.

Attempts to resolve merge conflicts and update modified/newly added
methods, and fixes the incorrect argument counts
on some of the Path methods.

### CODE REVIEW

**Comment by zanderso** [comment]
Looks like there's a segfault in a `Picture.toImageSync` unit test.

**Comment by dnfield** [comment]
I had an incorrect `FfiNative` annotation on the toImage sync and hadn't bothered running all the tests after getting stuff to compile on Friday afternoon. Canvas tests are passing now (I also was missing entries for a couple new methods on there for getting the transform and clip bounds). Let's see what CI thinks this time...

**Comment by dnfield** [comment]
CI is passing - this is ready for review.

**Comment by Hixie** [comment]
test-exempt: the existing tests should in theory cover this.

File: lib/ui/dart_ui.cc [hunk 1]
```diff
@@ -34,70 +34,278 @@
...
+  V(Path, extendWithPath, 4)                           \
+  V(Path, extendWithPathAndMatrix, 5)                  \
+  V(Path, getBounds, 1)                                \
+  V(Path, getFillType, 1)                              \
+  V(Path, lineTo, 3)                                   \
+  V(Path, moveTo, 3)                                   \
+  V(Path, op, 4)                                       \
+  V(Path, quadraticBezierTo, 5)                        \
+  V(Path, relativeArcToPoint, 8)                       \
+  V(Path, relativeConicTo, 6)                          \
```
Comment (dnfield) [inline lib/ui/dart_ui.cc:238]: I've fixed these compared to last time.

File: lib/ui/dart_ui.cc [hunk 2]
```diff
@@ -34,70 +34,278 @@
...
-    Vertices::RegisterNatives(g_natives);
-    PlatformConfiguration::RegisterNatives(g_natives);
+#define FFI_METHOD_MATCH(CLASS, METHOD, ARGS)                   \
+  if (strcmp(name, #CLASS "::" #METHOD) == 0 && args == ARGS) { \
+    return reinterpret_cast<void*>(                             \
+        tonic::FfiDispatcher<CLASS, decltype(&CLASS::METHOD),   \
+                             &CLASS::METHOD>::Call);            \
   }
+
+void* ResolveFfiNativeFunction(const char* name, uintptr_t args) {
```
Comment (zanderso) [inline lib/ui/dart_ui.cc:294]: Can this be a hash table lookup instead?

Comment (zanderso) [reply lib/ui/dart_ui.cc:294]: Also, does this get cached by the Dart VM, or do we pay this lookup cost on every call?

**Review by jason-simmons — APPROVED** [review]

File: lib/ui/compositing/scene_builder.cc [hunk 3]
```diff
@@ -32,44 +32,8 @@
...
-  V(SceneBuilder, addRetained)                      \
-  V(SceneBuilder, addPicture)                       \
-  V(SceneBuilder, addTexture)                       \
-  V(SceneBuilder, addPerformanceOverlay)            \
-  V(SceneBuilder, setRasterizerTracingThreshold)    \
-  V(SceneBuilder, setCheckerboardOffscreenLayers)   \
-  V(SceneBuilder, setCheckerboardRasterCacheImages) \
-  V(SceneBuilder, build)
-
-FOR_EACH_BINDING(DART_NATIVE_CALLBACK)
```
Comment (zanderso) [inline lib/ui/compositing/scene_builder.cc:65]: Does the `DART_NATIVE_CALLBACK` macro (and possibly other macros) become dead with this change? Can they be removed?

File: lib/ui/dart_ui.cc [hunk 4]
```diff
@@ -34,70 +34,278 @@
...
+// - Instantiate FfiDispatcher templates to automatically create FFI Native
+//   bindings.
+//   If the name does not match a native function, the template will fail to
+//   instatiate, resulting in a compile time error.
+// - Resolve the native function pointer associated with an @FfiNative function.
+//   If there is a mismatch between name or parameter count an @FfiNative is
+//   trying to resolve, an exception will be thrown.
+#define FFI_FUNCTION_LIST(V)                                          \
+  /* Constructors */                                                  \
+  V(Canvas::Create, 6)                                                \
```
Comment (zanderso) [inline lib/ui/dart_ui.cc:64]: This list should be alphabetized. But is there any way that these lists could go back into the separate files? That seemed easier to manage than this giant list.

File: lib/ui/dart_ui.cc [hunk 5]
```diff
@@ -34,70 +34,278 @@
...
+// These are used to:
+// - Instantiate FfiDispatcher templates to automatically create FFI Native
+//   bindings.
+//   If the name does not match a native function, the template will fail to
+//   instatiate, resulting in a compile time error.
+// - Resolve the native function pointer associated with an @FfiNative function.
+//   If there is a mismatch between names or parameter count an @FfiNative is
+//   trying to resolve, an exception will be thrown.
+#define FFI_METHOD_LIST(V)                             \
+  V(Image, dispose, 1)                                 \
```
Comment (zanderso) [inline lib/ui/dart_ui.cc:121]: Same comment about at least alphabetizing, and considering splitting back into separate files if possible.

File: lib/ui/compositing/scene_builder.cc (hunk 3 above)
Comment (dnfield) [reply lib/ui/compositing/scene_builder.cc:65]: They're still used in the Fuchsia embedding apparently.

File: lib/ui/dart_ui.cc (hunk 4 above)
Comment (dnfield) [reply lib/ui/dart_ui.cc:64]: They are alphabetized within their subsections (e.g. all constructors, all "other").

I don't have a strong preference either way here. This was part of Clement's original refactor.

File: lib/ui/dart_ui.cc (hunk 5 above)
Comment (dnfield) [reply lib/ui/dart_ui.cc:121]: I moved `Image` to the right section, everything else looked alphabetized.

I don't have a strong preference either way for whether we move these to separate files, but I'm somewhat inclined to get this landed first before it ends up with merge conflicts.

**Comment by dnfield** [comment]
Talked to Zach offline, going to try to land this and follow up on the nits.

File: lib/ui/dart_ui.cc (hunk 2 above)
Comment (jason-simmons) [reply lib/ui/dart_ui.cc:294]: I did a test implementation of map-based lookup a while ago while reviewing the original version of this PR.  I plan to revive it after this PR lands.

The lookup is cached by Dart and is only done once for each native function.

**Comment by dnfield** [comment]
Golem said this negatively impacted code size.

**Pull request merged**